eplus.py

```diff
@@ -37,31 +39,127 @@ def _get_eplus_data(session, eplus_url):
+                self._log.debug(f"Got new cookies: {repr(fresh_response.cookies)}")
+
+                # Filter cookies.
+                # For now, only the "ci_session" cookie is what we don't need, so ignore it.
+                cookie = next(
+                    cookie for cookie in fresh_response.cookies
+                        if cookie.name != "ci_session"
+                        and cookie.expires > time.time()
+                )
+                self._log.debug(
+                    "Found a valid cookie that will expire at "
+                    f"{time.strftime(r'%Y%m%d-%H%M%S%z', time.localtime(cookie.expires))}. "
+                    f"The cookie: {repr(cookie)}"
+                )
+
+                # Update the global session with the new cookies.
+                self._session.http.cookies.clear()
+                self._session.http.cookies.update(fresh_response.cookies)
+
+                self._retries = 0
+                self._last_expire_timestamp = cookie.expires
+
+                # Refresh cookies at most 5 minutes before expiration.
+                wait_sec = (cookie.expires - 5 * 60) - time.time()
+                if wait_sec < 0:
+                    # It's too close! Retry it right away.
+                    wait_sec = 0
+
+                self._log.debug(
+                    "Refreshed cookies. Next attempt will be at about "
+                    f"{time.strftime(r'%Y%m%d-%H%M%S%z', time.localtime(time.time() + wait_sec))}. "
+                )
+
+                self._closed.wait(wait_sec)
+                continue
+
+            except StopIteration as e:
+                # next() exhausted all cookies.
+                self._log.error("No valid cookies found.")
+
+            except Exception as e:
+                self._log.error(f"Failed to refresh cookies: {e}")
+
+            self._retries += 1
+            retry_delay_sec = 2 ** (self._retries - 1)
+
+            if time.time() + retry_delay_sec > self._last_expire_timestamp + 1 * 60 * 60:
+                self._log.error("We have not refreshed cookies in the past hour and will not try again.")
+
+                self.close()
+                return
+
+            self._log.debug(f"We will retry in {retry_delay_sec}s.")
+
+            self._closed.wait(retry_delay_sec)
+            continue
+
+    def _session_duplicator(self):
+        """
+        Make a duplicate of the member "_session" except for cookies.
+        """
+
+        new_session = HTTPSession()
+
+        new_session.proxies = self._session.http.proxies
+        new_session.headers = self._session.http.headers
+        new_session.trust_env = self._session.http.trust_env
+        new_session.verify = self._session.http.verify
+        new_session.cert = self._session.http.cert
+        new_session.timeout = self._session.http.timeout
+
+        return new_session
+
+
 class EplusHLSStreamWorker(HLSStreamWorker):
```

Comment:
This class can be removed entirely since the overridden `reload_playlist` is no longer needed. We can also drop the overridden `__worker__` behavior in `EplusHLSStreamReader`

Reply:
Ah, I'm going to add the feature "gracefully shutting down Streamlink if the playlist keeps unchanged for a while". The code will be placed in the "```reload_playlist```" function, so I preserved the "```EplusHLSStreamWorker```" class.

Reply:
The "```eplus_url```" argument in the constructor can be removed, I will commit it soon.

Reply:
> The "`eplus_url`" argument in the constructor can be removed, I will commit it soon.

Done in 2fd4a7ef85ae5ab676089d7133b6d35d2c919214 .